EASI-2513 - Add TRB Admin Notes

DylanSpOddball opened the pull request:
# EASI-2513

## Changes and Description

- Introduce TRB Admin Notes
    - Add SQL migration to create table for them
    - Create store methods for creating/fetching/updating them
        - These methods all follow the conventions we've established - they take a `context.Context` instead of just a logger, `GetTRBAdminNoteByID()` returns `nil, nil` if it doesn't find a matching note instead of logging an error (the resolver returns an error instead).
    - Add them to GraphQL schema
    - Add resolvers to link GQL operations with store methods


## Notes
* ~~I still want to add tests for the store methods, but I can get this up for review and running in CI first.~~ Tests added in https://github.com/CMSgov/easi-app/pull/1879/commits/4a35d68b27e75ccf0a31a8561352ed66719ff60c.
* I included `author: UserInfo!` (which includes EUA ID, common name, and email) as a field in the GraphQL type for `TRBAdminNote`, so the frontend doesn't have to make another request to look up the author's name based on their EUA ID.
* For handling archived (soft-deleted) notes, after a [discussion in Slack](https://cmsgov.slack.com/archives/CNU2B59UH/p1673376790722429), I decided to always return notes regardless of archived status for right now. Once the UX and requirements for archived notes are more developed and fully-formed (post-MVP), if we want/need to implement filtering on the backend, we can. The comments on the Jira ticket have some ideas for how to design this.
* Similar to https://github.com/CMSgov/easi-app/pull/1865 and https://github.com/CMSgov/easi-app/pull/1866, I didn't add any unit tests for the resolver functions due to their simplicity.
* ~~The tests added a utility function `createTRBRequestWithContext()`; if https://github.com/CMSgov/easi-app/pull/1877 is merged before this PR, that'll become redundant. We can move the existing `createTRBRequest()` function from `pkg/storage/trb_advice_letter_test.go` into `pkg/storage/store_test.go`, and use that function for this PR's tests as well.~~ Done as of https://github.com/CMSgov/easi-app/pull/1879/commits/db714aef395683c3ce33fb04d03951ef9e77b5fe.


## Questions for Reviewers
* Naming conventions - unlike most of the TRB-related models, this just uses a prefix of "TRB" instead of "TRBRequest" - it's `TRBAdminNote`/`trb_admin_note` instead of `TRBequestAdminNote`/`trb_request_admin_note`. If we want to change it to have the "Request" part of the prefix for consistency, we can. If we do that, we'll probably want to do the same for `TRBAdviceLetter`/`trb_advice_letter` instead of dropping the prefix.


## How to test this change

1. Spin up the backend and open the GraphQL playground.
2. Create a TRB request, noting the ID for later:
```gql
mutation {
  createTRBRequest(requestType: NEED_HELP) {
    id
    type
  }
}
```
3. Query for the notes on the request - there should be no notes (since none have been created), but no errors. (Save this query, it'll be used again)
```gql
query {
  trbRequest(id:"trbRequestIdFromAbove") {
    adminNotes {
      id
      trbRequestId
      isArchived
      category
      noteText
      author {
        euaUserId
        commonName
        email
      }
      createdBy
      createdAt
      modifiedBy
      modifiedAt
    }
  }
}
```
4. Create an admin note.
```gql
mutation createFirstNote {
  createTRBAdminNote(
    input: {
      trbRequestId: "trbRequestIdFromAbove"
      category: GENERAL_REQUEST
      noteText: "First note - Needs some more polish"
    }
  ) {
    id
    trbRequestId
    isArchived
    category
    noteText
    author {
      euaUserId
      commonName
      email
    }
    createdBy
    createdAt
    modifiedBy
    modifiedAt
  }
}
```
5. Run the query from step 3 again. You should see the created note, with the correct category and text, and `isArchived` set to `false`.
6. Create a second admin note.
```gql
mutation createSecondNote {
  createTRBAdminNote(
    input: {
      trbRequestId: "trbRequestIdFromAbove"
      category: ADVICE_LETTER
      noteText: "Second note - Maybe soften the criticism a bit"
    }
  ) {
    id
    trbRequestId
    isArchived
    category
    noteText
    author {
      euaUserId
      commonName
      email
    }
    createdBy
    createdAt
    modifiedBy
    modifiedAt
  }
}
```
7. Run the query from step 3 again. You should see both created notes, with distinct data for each note.
8.  Update the first admin note, changing the local authentication header in the GraphQL playground to use `TEST` as the EUA ID, using the note ID returned from creating the note in step 4.
```gql
mutation updateFirstNote {
  updateTRBAdminNote(
    input: {
      id: "trbAdminNoteFromStep4"
      category: INITIAL_REQUEST_FORM
      noteText: "First note - updated"
    }
  ) {
    id
    trbRequestId
    isArchived
    category
    noteText
    author {
      euaUserId
      commonName
      email
    }
    createdBy
    createdAt
    modifiedBy
    modifiedAt
  }
}
```
9. Run the query from step 3 again. The first note should show the updated values for the `category` and `noteText` fields, the `modifiedBy` field should be `"TEST"`, and `modifiedAt` should show the time when the update mutation was ran.
10. Archive the second admin note, changing the local authentication header in the GraphQL playground to use `GRTB` as the EUA ID, using the ID returned from creating the note in step 6.
```gql
mutation archiveSecondNote {
  archiveTRBAdminNote(id: "trbAdminNoteIdFromStep6") {
    id
    trbRequestId
    isArchived
    category
    noteText
    author {
      euaUserId
      commonName
      email
    }
    createdBy
    createdAt
    modifiedBy
    modifiedAt
  }
}

```
11. Run the query from step 3 again. The second note should show that `isArchived` is now `true`, the `modifiedBy` field should be `"GRTB"`, and `modifiedAt` should show the time when the archival mutation was ran.

ClayBenson94 commented:
@DylanSpOddball I just approved https://github.com/CMSgov/easi-app/pull/1877 - If you want to merge that, I think it would be good to be able to nix the `createTRBRequestWithContext()` function, as you mention in the notes!

ClayBenson94 reviewed (commented):
> Naming conventions - unlike most of the TRB-related models, this just uses a prefix of "TRB" instead of "TRBRequest" - it's TRBAdminNote/trb_admin_note instead of TRBequestAdminNote/trb_request_admin_note. If we want to change it to have the "Request" part of the prefix for consistency, we can. If we do that, we'll probably want to do the same for TRBAdviceLetter/trb_advice_letter instead of dropping the prefix.

I think this is fine. I don't think it's confusing that a TRB Admin Note (or TRB Advice Letter) is associated with a request, and the shorter naming makes reading a bit easier - I'm alright with this naming!

DylanSpOddball commented:
> @DylanSpOddball I just approved #1877 - If you want to merge that, I think it would be good to be able to nix the `createTRBRequestWithContext()` function, as you mention in the notes!

Done in https://github.com/CMSgov/easi-app/pull/1879/commits/db714aef395683c3ce33fb04d03951ef9e77b5fe.

jstrang-oddball commented on pkg/graph/resolvers/trb_admin_note.go:
```diff
@@ -0,0 +1,106 @@
+
+// GetTRBAdminNoteAuthorInfo gets the full user info for the author of an admin note
+func GetTRBAdminNoteAuthorInfo(ctx context.Context, authorEUAID string, fetchUserInfo func(context.Context, string) (*models.UserInfo, error)) (*models.UserInfo, error) {
+	authorInfo, err := fetchUserInfo(ctx, authorEUAID)
```
This seems like a case where we could skip having a function in the resolvers package, since all you're doing is passing a function into the resolvers package so it can call that function and return the results.

DylanSpOddball replied:
Yeah, I was on the fence about separating it out when I wrote it; inlining it in `schema.resolvers.go` probably makes more sense.

DylanSpOddball replied:
Inlined in https://github.com/CMSgov/easi-app/pull/1879/commits/3a602d1344bc57d2597513bfda620580ba0b3ecb.

DylanSpOddball commented:
Feedback from a brief PR review with Clay and Jeremy:
* Remove the possibility of setting `isArchived` through the `updateTRBAdminNote` GQL mutation.
* Change the `archiveTRBAdminNote` mutation to a more general `setTRBAdminNoteArchived` mutation that can set archived to true/false.
* put the `@hasRole(role: EASI_TRB_ADMIN)` directive on:
    * `trbAdminNote` query
    * The `adminNotes` field on the TRBRequest type
    * The new `setTRBAdminNoteArchived` mutation.

DylanSpOddball commented:
> * Remove the possibility of setting `isArchived` through the `updateTRBAdminNote` GQL mutation.
> * Change the `archiveTRBAdminNote` mutation to a more general `setTRBAdminNoteArchived` mutation that can set archived to true/false.

Implemented in https://github.com/CMSgov/easi-app/pull/1879/commits/7259b8562195aa08a81f0d6eb5e26a48d7afadbf.

> * Put the `@hasRole(role: EASI_TRB_ADMIN)` directive on:
>   
>   * `trbAdminNote` query
>   * The `adminNotes` field on the TRBRequest type
>   * The new `setTRBAdminNoteArchived` mutation.

Implemented in https://github.com/CMSgov/easi-app/pull/1879/commits/6f9394a0f0a6531a7e8c9c06294aed1b0b39613c.

@jstrang-oddball @ClayBenson94 , do you want me to edit the description/add a post with tests for the changes? They're pretty straightforward to test.

DylanSpOddball commented:
One note about putting `@hasRole(role: EASI_TRB_ADMIN)` on `TRBRequest`'s `adminNotes` field - any requests without the job code that include the `adminNotes` field will _entirely_ fail as unauthorized, they won't return any data for fields that don't have the directive on them. Example:
```gql
query {
  trbRequest(id:"someTRBRequestID") {
    name
    status
    adminNotes {
      isArchived
      category
      noteText
    }
  }
}
```

Requests without the job code that don't ask for `adminNotes` will complete successfully. Example:
```gql
query {
  trbRequest(id:"someTRBRequestID") {
    name
    status
  }
}
```

Is this the desired behavior, specifically that requesting an unauthorized field errors entirely without returning any data?

jstrang-oddball reviewed (approved):
Looks good!

ClayBenson94 commented:
> One note about putting `@hasRole(role: EASI_TRB_ADMIN)` on `TRBRequest`'s `adminNotes` field - any requests without the job code that include the `adminNotes` field will _entirely_ fail as unauthorized, they won't return any data for fields that don't have the directive on them. Example:
> 
> ```graphql
> query {
>   trbRequest(id:"someTRBRequestID") {
>     name
>     status
>     adminNotes {
>       isArchived
>       category
>       noteText
>     }
>   }
> }
> ```
> 
> Requests without the job code that don't ask for `adminNotes` will complete successfully. Example:
> 
> ```graphql
> query {
>   trbRequest(id:"someTRBRequestID") {
>     name
>     status
>   }
> }
> ```
> 
> Is this the desired behavior, specifically that requesting an unauthorized field errors entirely without returning any data?

I think this makes sense. It'd be interesting to see if only returning an error for that property is possible, but solely out of curiosity. I actually think I prefer the whole thing erroring because it makes it clear that the request was faulty by nature, rather than a small piece of the request breaking.
